### PR TITLE
feat!: replace all bbtool wrappers with one bbtool wrapper for all subcommands

### PR DESCRIPTION
**Pull request by SilasK**
### Description

I add a generic bbtool wrapper. 
It is based on #2 

BBtools hase many bash shcripts. This wrapper should allow to run them all. 

### QC

* [x] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

### CODE REVIEW

File: bio/bbtools/generic/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    logfile = get_stderr_file(snakemake)
+
+    # clear stderr file
+    open(logfile, "w").close()
+
+except:
+    print("No log file provided, logging to stdout")
+    logfile = None
+
+logging.basicConfig(
```
Comment (SilasK) [inline bio/bbtools/generic/wrapper.py:83]: That's my way to enshure that all errors in the wrapper are written to the log file

File: bio/bbtools/generic/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    )
+
+
+# Install exception handler
+sys.excepthook = handle_exception
+
+logger = logging.getLogger(__file__)
+
+##############################################################
+## helper functions they should go in snakemake_wrapper_utils
```
Comment (SilasK) [inline bio/bbtools/generic/wrapper.py:112]: helper functions they should go in snakemake_wrapper_utils

File: bio/bbtools/generic/wrapper.py (hunk 1 above)
Comment (SilasK) [reply bio/bbtools/generic/wrapper.py:83]: #885

**Comment by SilasK** [comment]
Can I ask @fgvieira  and or @dlaehnemann for a review. 

@fgvieira You did the other bbtools snakemake wrappers, isn't it?

File: bio/bbtools/generic/wrapper.py (hunk 2 above)
Comment (SilasK) [reply bio/bbtools/generic/wrapper.py:121]: @dlaehnemann Do I need to do this first?

File: bio/bbtools/generic/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+input:
+  - input: All keywords in input are passed to the bbtool as key=value pairs. A special case is 'input' see in the notes. 
+  - flag: The keyword 'flag' is ignored and allows to make input files that are not processed by the bbtool.
+output:
+  - out: All keywords in output are passed to the bbtool as key=value pairs. A special case is 'out' see in the notes. 
+params:
+  - command: Required parameter defining the command to be used e.g. 'bbmap.sh'
+  - extra: additional program arguments
+  - other: All other parameters are passed to the bbtool as key=value pairs
+  - define_threads: By default the the threads argument of snakemake is passed to the bbtool. For bbtools that do not support this argument, set this to False.
```
Comment (fgvieira) [inline bio/bbtools/generic/meta.yaml:16]: Instead of this `param`, what if there was a list of all the BBduk tools that support threads in the wrapper?

File: bio/bbtools/generic/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  - Silas Kieser
+input:
+  - input: All keywords in input are passed to the bbtool as key=value pairs. A special case is 'input' see in the notes. 
+  - flag: The keyword 'flag' is ignored and allows to make input files that are not processed by the bbtool.
+output:
+  - out: All keywords in output are passed to the bbtool as key=value pairs. A special case is 'out' see in the notes. 
+params:
+  - command: Required parameter defining the command to be used e.g. 'bbmap.sh'
+  - extra: additional program arguments
+  - other: All other parameters are passed to the bbtool as key=value pairs
```
Comment (fgvieira) [inline bio/bbtools/generic/meta.yaml:15]: It seems this `param` is not used.

**Comment by fgvieira** [comment]
Nice PR! Can you also add tests for the other rules on the `test/Snakefile`?

And maybe even add the tests from the other `bbduk` wrappers (`bbmap`, `tadpole`, and `loglog`) since, if this wrapper works with all, I'd even say to delete the other wrappers and have this one as the single wrapper for `bbduk`.

File: bio/bbtools/generic/environment.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+channels:
+  - conda-forge
+  - bioconda
+  - nodefaults
+dependencies:
+  - bbmap =39.01
+  - snakemake-wrapper-utils =0.5
+  - python =3.11
```
Comment (fgvieira) [inline bio/bbtools/generic/environment.yaml:8]: Is python really necessary?

Comment (fgvieira) [reply bio/bbtools/generic/environment.yaml:8]: I tested the other wrappers (and made #1366), and they seem to run fine without it.

Comment (johanneskoester) [reply bio/bbtools/generic/environment.yaml:8]: Yes, if there is no python in the env, Snakemake uses the python from its main process.

**Review by johanneskoester — CHANGES_REQUESTED** [review]
I also think the other bbtool wrappers should be deleted in order to avoid duplication. One should probably just call the wrapper bbtools, without the "generic" then.

**Comment by fgvieira** [comment]
And all add tests for other bbtool tools, e.g. `clumpify`.

File: bio/bbtools/generic/wrapper.py (hunk 2 above)
Comment (SilasK) [reply bio/bbtools/generic/wrapper.py]: Finally now as this will be the only wrapper that uses this

**Review by SilasK — COMMENTED** [review]
I moved them

**Review by SilasK — COMMENTED** [review]
I moved the wrapper to the base dir `bio/bbtools` and removed the others. I combined all the tests

File: bio/bbtools/environment.yaml [hunk 6]
```diff
@@ -4,4 +4,5 @@ channels:
   - nodefaults
 dependencies:
   - bbmap =39.01
-  - snakemake-wrapper-utils =0.5.3
+  - snakemake-wrapper-utils =0.5
```
Comment (fgvieira) [inline bio/bbtools/environment.yaml:7]:
```suggestion
  - snakemake-wrapper-utils =0.6.2
```

File: bio/bbtools/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+input:
+  - input: All keywords in input are passed to the bbtool as key=value pairs. A special case is 'input' see in the notes.
+  - flag: The keyword 'flag' is ignored and allows to make input files that are not processed by the bbtool. All other input are passed to the bbtool as key=value pairs
+output:
+  - out: All keywords in output are passed to the bbtool as key=value pairs. A special case is 'out' see in the notes. All other outputs are passed to the bbtool as key=value pairs
+params:
+  - command: Required parameter defining the command to be used e.g. 'bbmap.sh'
+  - extra: additional program arguments. All other parameters are passed to the bbtool as key=value pairs
+notes: |
+  This wrapper allows to run any of the bbtools. The command is defined by the 'command' parameter, which required.
```
Comment (fgvieira) [inline bio/bbtools/meta.yaml:16]:
```suggestion
  This wrapper allows to run any of the bbtools. The command is defined by the 'command' parameter, which is required.
```

File: bio/bbtools/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,133 @@
+name: Generic
+url: https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide
```
Comment (fgvieira) [inline bio/bbtools/meta.yaml:2]:
```suggestion
url: https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/
```

File: bio/bbtools/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  **All keywords in input, output, params are passed as key value pairs to the command!**
+  Take care that they are valid for the bbtool.
+
+  As it it not possible to define 'in' as a keyword, the keyword 'input' is used instead. Allowed aliases are 'sample' and 'reads', but we recommend to use 'input'.
+  If exactly two files are provides for 'input' or 'out' the wrapper parses them as in1, in2 or out1, out2, respectively.
+
+  For all parameters, if more than two files are provided, the wrapper parses them as key=value1,value2,value3... 
+
+  Limitations:
+    - The wrapper cannot run bbwarap with paired + singleton files. But this is a rare use case.
```
Comment (fgvieira) [inline bio/bbtools/meta.yaml:26]: What is the `bbwarap` tool? I could not find it on the documentation.

**Comment by fgvieira** [comment]
Left some minor comments but, otherwise, great work!

File: bio/bbtools/meta.yaml (hunk 9 above)
Comment (SilasK) [reply bio/bbtools/meta.yaml]: bbwrap can process single + paired end reads, which uses some special input format for in1, in2.

**Comment by SilasK** [comment]
@fgvieira  Do we need to wait for @johanneskoester  or can you accept the PR.

There will be probably some edge cases in the future. but I think it's quite neat for now.

File: bio/bbtools/test/Snakefile [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+rule index_genome:
+    input:
+        ref="genome.fasta",
+    output:
+        flag=touch("genome.indexed"),
+    params:
+        command="randomreads.sh",
+        build= 1
+    log:
+        "logs/index_genome/{sample}.log",
```
Comment (fgvieira) [inline bio/bbtools/test/Snakefile:56]:
```suggestion
        "logs/index/genome.log",
```

**Comment by SilasK** [comment]
Cool all checks have passed!

**Comment by SilasK** [comment]
I think @johanneskoester You need to approve your requested changes

File: bio/bbtools/environment.yaml (hunk 6 above)
Comment (fgvieira) [reply bio/bbtools/environment.yaml:7]: Any reason for not using the latest `snakemake_wrapper_utils` version?

**Comment by fgvieira** [comment]
It seems there is an issue with the test files.

**Comment by SilasK** [comment]
It worked before 😠 .
It seems a problem with, writing and reading to the index.


BBmap stores index files in `ref/{buildnr}` it allows to change the build number but not really to change the path.
One can ignore the path and set different build numbers if needed. 
But I now set the index directory as an output, which makes things more explicit.

But now, it doesn't really correspond with the wrapper's guideline that input and output can be freely set.

I hope you grant this exception.

**Comment by SilasK** [comment]
It is passing the checks now @fgvieira Thank you for your patience..

File: bio/bbtools/test/Snakefile [hunk 11]
```diff
@@ -44,14 +46,15 @@ rule pileup:  # all rules use the same wrapper!!
...
 
 
+
 rule index_genome:
     input:
         ref="genome.fasta",
     output:
-        flag=touch("genome.indexed"),
+        # Unfortunately bbmap requires an index to be stored in this folder
+        flag=temp(directory("ref/genome/1"))
```
Comment (fgvieira) [inline bio/bbtools/test/Snakefile]: Should it be a flag, or can it be a `multiext`?

Comment (fgvieira) [reply bio/bbtools/test/Snakefile]: Just made a quick test, and it seems the `bbmap` index is a variable number of output files (with variable names). Isn't that so?

**Comment by dlaehnemann** [comment]
> It worked before 😠 . It seems a problem with, writing and reading to the index.
> 
> BBmap stores index files in `ref/{buildnr}` it allows to change the build number but not really to change the path. One can ignore the path and set different build numbers if needed. But I now set the index directory as an output, which makes things more explicit.
> 
> But now, it doesn't really correspond with the wrapper's guideline that input and output can be freely set.
> 
> I hope you grant this exception.

I think you can actually give it an alternative path using the `path=/another/location/` command line argument, which would then be the way to go here. See point `8.` for the index creation here:
https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/

**Comment by SilasK** [comment]
so bbmap index, creates files in `ref/genome/{build_nr}` and `ref/index/{buildnr}`
with `path=path/to /folder` we can give a folder where the ref folder is placed. But, it doesn't correspond to the goal of freely setting an output path.

**Comment by dlaehnemann** [comment]
But it seems like you are already trying things along those lines... :D

**Comment by SilasK** [comment]
Thank you for the suggestion. It should work now.

**Comment by fgvieira** [comment]
Can you bump the `snakemake-wrappers` version?

**Comment by dlaehnemann** [comment]
> Thank you for the suggestion. It should work now.

And it was easier than I thought. Cool!

**Comment by SilasK** [comment]
version updated

**Comment by SilasK** [comment]
But apparently, bbmap's jave doesn't understand the `-Djava.io.tmpdir=/tmp` 

Should I deowngread the snakemake_wrapper_utils

**Comment by fgvieira** [comment]
Damn... :disappointed:
Not ideal, but I guess we can pin `snakemake-wrapper-utils` to version `0.5.3` until `bbtools` supports it.
@dlaehnemann @johanneskoester What do you think?

**Comment by dlaehnemann** [comment]
Arrgh, that's ugly. None of the BBTools shell scripts passes anything except for the memory specification on to Java. I guess the thorough (but laborious) fix would be to switch to the actual java script syntax, namely this:
```


 Maybe we should add an argument to `get_java_opts(adjust_tmpdir=True)`, which you can then unset in your `get_java_opts()` call.

**Comment by dlaehnemann** [comment]
I think you could parse the file `ToolDescriptions.txt` from the folder `bbmap/docs/` of the [downloadable `BBTools` source code from sourceforge](https://sourceforge.net/projects/bbmap/files/latest/download) to get a mapping from shell script to java tool. But then I also just saw that they seem to pass some hard-coded default values to some tools. So switching from the shell scripts will loose default parameter settings... :exploding_head:

**Comment by dlaehnemann** [comment]
Hmm, but if we look at the actual java command that is executed, the option is still in there. So it is simply the sanity checking of BBTools that is off here. Maybe passing `-da` will turn this off? But then, this might turn other useful checks off...

**Comment by dlaehnemann** [comment]
Mhmmm, no. It seems like `-da` will not turn off the failure we get...

**Comment by SilasK** [comment]
Alternatively, we could re-implement the get_java_opts inside the wrapper?

**Comment by dlaehnemann** [comment]
Or remove the respective part from the resulting `java_opts` string via a regex. :see_no_evil:

**Comment by dlaehnemann** [comment]
Or maybe we can change the `get_java_opts` to instead set all of the general option in `os.environ["JDK_JAVA_OPTIONS"] = "-Djava.io.tmpdir={}"`. This sounds like this should then automatically use those options in any following `java` call:
https://docs.oracle.com/en/java/javase/17/docs/specs/man/java.html#using-the-jdk_java_options-launcher-environment-variable

Why are Java tools always such a headache... :disappointed:

**Comment by SilasK** [comment]
Change impelemented.

**Comment by dlaehnemann** [comment]
Nice. But yes, (not) knowing Java and all, I am also a bit concerned this might not work as expected. And have no idea how to easily test this.

Maybe one could try to locally (and manually) run the tests on this pull request and on some other wrapper(s) that use the `get_java_opts()` function, while having manually modified the respective conda environments' `snakemake-wrapper-utils` to include the change...?!

**Comment by dlaehnemann** [comment]
I think you might have to split the `'--default-resources mem_mb=200'` into two separate arguments, so `'--default-resources', 'mem_mb=200'`.

**Comment by SilasK** [comment]
@dlaehnemann  To follow up on 
> But this is something that the user might want to set via snakemake command-line options or via the `resources: tempdir=` directive in a rule using the wrapper, so it would then be good if the wrapper did pass this on to the actual execution of the tool.
https://github.com/snakemake/snakemake-wrapper-utils/pull/37#issuecomment-1805909767

I don't think bbmap uses tempdirs a lot. But I found that one command can use it.

> The location of temp files can be specified with the “tmpdir” flag; it defaults to the environment variable $TMPDIR, which on Genepool points to local disk when available. Temp files will be cleaned up once BBNorm finishes.

`$TMPDIR` will already be set by snakemake using the `resources.tempdir` isn't it. So I would say the specification of tempdir should be ok, fo rbbtools.

File: bio/bbtools/wrapper.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+    # Getting memory.
+    if "-Xmx" in extra:
+        sys.exit(java_mem_xmx_error("extra"))
+
+    mem_mb = get_mem(snakemake)
+    logger.debug(
+        f"Memory specified {mem_mb} MiB, use {(1.0 - java_mem_overhead_factor)*100}% of it for Java heap."
+    )
+
+    return " -Xmx{}M".format(round(mem_mb * (1.0 - java_mem_overhead_factor)))
```
Comment (dlaehnemann) [inline bio/bbtools/wrapper.py:221]: Sorry to have one more suggestion here: wouldn't it be a bit more future-proof to simply use the snakemake-wrapper-utils `get_java_opts` and simply extract only the `-Xmx` option with a regex? That way, we don't reimplement it here and if something gets fixed in the utils function, we also benefit here.

**Review by dlaehnemann — COMMENTED** [review]
It seems like I had some comments a while ago, that are still lurking around. I think only the latest one is relevant and will try to resolve all the others. Sorry for the noise.

And many thanks for following up on the `$TMPDIR` question. If `bbtools` don't seem to use it and if they do, to respect the environment variable, then I think you are right -- we can ignore this for the wrapper...

**Comment by SilasK** [comment]
I now extract the mem using a regex.

I also made a major update to the Snakefile test. It now represents a good usage of the wrapper and the bbtools functions. 
e.g. previously all parameters were input as a comma-separated string. 
With the same rules, it works for paired or single-end reads.

What do you think of my updated workflow?

**Comment by fgvieira** [comment]
I think it looks good but, since it will introduce breaking changes, not sure if @johanneskoester want to take a look first.

**Review by dlaehnemann — APPROVED** [review]
This generally looks good to me. But I haven't looked at all the wrapper.py code in detail -- @johanneskoester has requested changes anyway, and will thus have to have another look. I'll ping him...

**Review by johanneskoester — APPROVED** [review]

File: bio/bbtools/test/Snakefile [hunk 13]
```diff
@@ -0,0 +1,188 @@
+include: "input_functions.smk"
```
[the remark this answers is not in the document]
Comment (SilasK) [reply bio/bbtools/test/Snakefile:1]: I had it this way. But then the lining didn't pass.

File: bio/bbtools/test/Snakefile [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        expand("logs/loglog/{sample}.log", sample=["SampleA"]),
+        expand("covstats/{sample}.tsv", sample=["SampleA"]),
+
+
+rule index_genome:
+    input:
+        ref="genome.fasta",
+    output:
+        # path to the index base folder, The index is stored in {path}/ref/genome/{build} and {path}/ref/index/{build}
+        # With build by defualt 1 and path by default the working directory
```
[the remark this answers is not in the document]
Comment (SilasK) [reply bio/bbtools/test/Snakefile:15]: Maybe it's not clear, The comment explains the workings of path and build. 
But maybe we don't need to bother the users with that.

[the remark this answers is not in the document]
Comment (SilasK) [reply bio/bbtools/test/Snakefile]: I adapted the commend and moved it down in the Snakefile as it is less important.

File: bio/bbtools/wrapper.py [hunk 15]
```diff
@@ -0,0 +1,456 @@
...
+
+# global flags to check if input and output are parsed multiple times
+parsed_input, parsed_output = False, False
+
+import warnings
+from snakemake_wrapper_utils import java
+
+
+import sys
+import warnings
```
[the remark this answers is not in the document]
Comment (SilasK) [reply bio/bbtools/wrapper.py]: imported at the beginning now.

I have a second section of imports after all the logging is set up.
So that errors in imports would also be logged to the log file.

**Comment by SilasK** [comment]
Now we are almost done!

I had two problems with the linter:

1. It didn't want me to have functions in the main Snakemake, therefore the import. Even if I think it's less ideal.

2. It interpreted a `ktrim="r"` as a path, and didn't accept it. I found a solution, which is less optimal.

**Comment by SilasK** [comment]
In the wrapper I put the links to other PRs in snakemake that will incorporate these functions. But I guess we should not wait until then to release the next version of this wrapper.

File: bio/bbtools/test/Snakefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (dlaehnemann) [reply bio/bbtools/test/Snakefile:1]: Ah, I see. That's a bit annoying in this setup. But I guess there's no way around it for now. Maybe add a quick comment above the include, which rule this is for and the function name? Although the file name is rather descriptive already... :D

File: bio/bbtools/wrapper.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (dlaehnemann) [reply bio/bbtools/wrapper.py]: Ah, OK. That makes sense. Maybe add a quick comment at the start, that there is another import block below, at the `start of the wrapper`?

**Comment by dlaehnemann** [comment]
> Now we are almost done!
Yes, definitely!

> I had two problems with the linter: 
>     1. It didn't want me to have functions in the main Snakemake, therefore the import. Even if I think it's less ideal.
Then I guess we'll leave it as is. Maybe add a comment as suggested, if you think this is useful.
 
>     2. It interpreted a `ktrim="r"` as a path, and didn't accept it. I found a solution, which is less optimal.
Is that under params? Because I would find that weird, and possibly a linter bug. But if it enforces this for input and output, this makes sense. Then, I would suggest simply specifying this as a param. And if that's what you did, did you try something like either of the following, to make it work with the current linter:
```
params:
    ktrim="'r'",
    ktrim="'r' ",
    ktrim="r ",
```

As for the links to PRs, I think it is perfectly fine to simply leave those as TODOs. Maybe make it a bit clearer still, that once these are merged, the respective functions can simply go away and use snakemake internal functionality.

**Comment by SilasK** [comment]
It looks like a linter bug:
[Code](https://github.com/metagenome-atlas/snakemake-wrappers/blob/37d0bf2dfcfbde20269090c687e4184136033779/bio/bbtools/test/Snakefile#L88C1-L92C19)
[Error](https://github.com/snakemake/snakemake-wrappers/actions/runs/6849925541/job/18623086042)

**Comment by dlaehnemann** [comment]
Ah, thanks for the pointers to the linter. Regarding the (input) function, you could indeed make this an inline lambda expression like this:

```
rule loglog:
    input:
        expand(
            "reads/raw/{{sample}}_{fraction}.fastq.gz",
            fraction=lambda wc: ["R1", "R2"] if config.get("reads_are_paired", True) else ["se"]
        ),
```

It has the advantage of being right there in the input of the rule. It has the disadvantage of being slightly less readable than the actual function with newlines. And you'll have to repeat that code wherever you provide such a read `fraction` specification. It seems like there's no perfect solution with the current linter.

And from what I understand about the `ktrim="r"`, this recognizes that `r` is a prefix of the input/output path, which starts with `reads/...`. That's a very short prefix, but it is correct. The linter could probably be amended to ignore such single letter cases or make it more of a suggestion, then a must (not sure if that exists, though). And I think you should get around this by including some random whitespace at the end, so:
`ktrim="r "`
Then, it shouldn't be a prefix any more...

**Comment by SilasK** [comment]
The whitespace fixed the linter.

I find the lambda version less clear.

**Comment by dlaehnemann** [comment]
Agreed, I'll merge! :tada:

**Comment by SilasK** [comment]
Thank you very much for your review and your suggestions!

**Comment by dlaehnemann** [comment]
Thanks for keeping at it! Getting things right for reproducible and reusable data analyses can be exhausting, especially when you're battling with multiple levels of the workflow / analysis stack and several workflows at the same time.

So take a moment and celebrate! :rocket:

**Comment by fgvieira** [comment]
Nice work!!
